components/datetime/src/metazone.rs

```diff
@@ -0,0 +1,121 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::provider::time_zones::{MetaZoneId, TimeZoneBcp47Id};
+
+use crate::error::DateTimeFormatterError;
+use crate::provider::time_zones::MetaZonePeriodV1Marker;
+use icu_calendar::DateTime;
+use icu_calendar::Iso;
+use icu_locid::Locale;
+use icu_provider::prelude::*;
+use zerovec::ule::AsULE;
+
+/// [`MetaZoneCalculator`] uses data from the [data provider], the selected [`Locale`] to calculate
+/// metazone id.
+pub struct MetaZoneCalculator {
+    pub(super) metazone_period: DataPayload<MetaZonePeriodV1Marker>,
+}
+
+impl MetaZoneCalculator {
+    /// Constructor that loads data before calculating metazone id.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_locid::locale;
+    /// use icu::datetime::metazone::MetaZoneCalculator;
+    ///
+    /// let provider = icu_testdata::get_provider();
+    /// let mzc = MetaZoneCalculator::new(
+    ///     locale!("en"),
+    ///     &provider,
+    /// );
+    ///
+    /// assert!(mzc.is_ok());
+    /// ```
+    pub fn new<L, ZP>(locale: L, zone_provider: &ZP) -> Result<Self, DateTimeFormatterError>
+    where
+        L: Into<Locale>,
+        ZP: DataProvider<MetaZonePeriodV1Marker> + ?Sized,
+    {
+        let locale = locale.into();
```

Comment:
I think I've mentioned this before, but I think this data should be in a single `und` locale. You shouldn't need to take the locale via the API because all locales have the same data. You should collapse the data into a single `und.json` file in the datagen side of this code.

```
sffc@sffc:~/projects/icu4x/provider/testdata/data/json/time_zone/metazone_period@1$ shasum *
b4f1eb5b621f3de2db6df3876385a74ad705ed50  ar-EG.json
b4f1eb5b621f3de2db6df3876385a74ad705ed50  ar.json
b4f1eb5b621f3de2db6df3876385a74ad705ed50  bn.json
b4f1eb5b621f3de2db6df3876385a74ad705ed50  ccp.json
b4f1eb5b621f3de2db6df3876385a74ad705ed50  en-001.json
b4f1eb5b621f3de2db6df3876385a74ad705ed50  en.json
b4f1eb5b621f3de2db6df3876385a74ad705ed50  en-ZA.json
b4f1eb5b621f3de2db6df3876385a74ad705ed50  es-AR.json
b4f1eb5b621f3de2db6df3876385a74ad705ed50  es.json
b4f1eb5b621f3de2db6df3876385a74ad705ed50  fil.json
b4f1eb5b621f3de2db6df3876385a74ad705ed50  fr.json
b4f1eb5b621f3de2db6df3876385a74ad705ed50  ja.json
b4f1eb5b621f3de2db6df3876385a74ad705ed50  ru.json
b4f1eb5b621f3de2db6df3876385a74ad705ed50  sr-Cyrl.json
b4f1eb5b621f3de2db6df3876385a74ad705ed50  sr.json
b4f1eb5b621f3de2db6df3876385a74ad705ed50  sr-Latn.json
b4f1eb5b621f3de2db6df3876385a74ad705ed50  th.json
b4f1eb5b621f3de2db6df3876385a74ad705ed50  tr.json
b4f1eb5b621f3de2db6df3876385a74ad705ed50  und.json
```

Reply:
Track in https://github.com/unicode-org/icu4x/issues/2257